ocis-pkg/runner/runner.go

```diff
@@ -0,0 +1,104 @@
+package runner
+
+import (
+	"context"
+)
+
+// Runner represents the one executing a long running task, such as a server
+// or a service.
+// The ID of the runner is public to make identification easier, and the
+// Result that it will generated will contain the same ID, so we can
+// know which runner provided which result.
+type Runner struct {
+	ID        string
+	fn        Runable
+	interrupt Stopper
+}
+
+// NewRunner will create a new runner.
+// The runner will be created with the provided id (the id must be unique,
+// otherwise undefined behavior might occur), and will run the provided
+// runable task, using the "interrupt" function to stop that task if needed.
+//
+// Note that it's your responsibility to provide a proper stopper for the task.
+// The runner will just call that method assuming it will be enough to
+// eventually stop the task at some point.
+func NewRunner(id string, fn Runable, interrupt Stopper) *Runner {
+	return &Runner{
+		ID:        id,
+		fn:        fn,
+		interrupt: interrupt,
+	}
+}
+
+// Run will execute the task associated to this runner in a synchronous way.
+// The task will be spawned in a new goroutine, and the current thread will
+// wait until the task finishes.
+//
+// The task will finish "naturally". The stopper will be called in the
+// following ways:
+// - Manually calling this runner's `Interrupt` method
+// - When the provided context is done
+// As said, it's expected that calling the provided stopper will be enough to
+// make the task to eventually complete.
+//
+// Once the task finishes, the result will be returned.
+//
+// Some nice things you can do:
+// - Use signal.NotifyContext(...) to call the stopper and provide a clean
+// shutdown procedure when an OS signal is received
+// - Use context.WithDeadline(...) or context.WithTimeout(...) to run the task
+// for a limited time
+func (r *Runner) Run(ctx context.Context) *Result {
+	ch := make(chan *Result)
+
+	go func(ch chan<- *Result) {
+		err := r.fn()
+
+		result := &Result{
+			RunnerID:    r.ID,
+			RunnerError: err,
+		}
+		ch <- result
+		close(ch)
+	}(ch)
+
+	select {
+	case result := <-ch:
+		return result
+	case <-ctx.Done():
+		r.interrupt()
+	}
+
+	return <-ch
```

Comment:
This will block forever when stopper is misconfigured. We need some way of emergency exit here.

Reply:
I'm not sure if we can do anything about it. Without knowing anything about the task, we'll have to rely on the stopper to do its job. The alternative would be to wait until the task finishes naturally.

I'll check if our servers provide something that could lead to a "natural" shutdown other than calling the server's `Stop` method. I'd assume that having the provided context marked as done would make the server to shutdown.

Reply:
I think it's still pretty task-dependent. You can still create a task with a timeout to ensure the task finishes in 1 hour, for example.

It might be better to create a "helpers.go" file to handle these things. For example `TimedTask(r *Runable, ch chan error, dur time.Duration) (Runable, Stopper)` in order to wrap the target task and get a suitable task with its stopper.
This way, the core functionality is simple, and the complexity is moved away.

In any case, this would only help to unblock the thread. The underlying goroutine (the GRPC server, for example, in our case) would keep running and eating resources if no proper stopper have been supplied.

Reply:
move to line 70 please